File: android/app/src/main/kotlin/com/cacompte/app/ui/GameRequestMail.kt
package com.cacompte.app.ui

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import com.cacompte.app.R

/**
 * Doc utilisateur — demande d'ajout d'un jeu : pas de formulaire ni de backend dédié, un e-mail
 * préempli suffit. Miroir de `GameRequestMail.swift`, adapté à l'idiome Android (`ACTION_SENDTO`
 * + extras plutôt qu'une URL `mailto:` reconstruite à la main). `context.getString(...)`, pas
 * `stringResource()` (Compose) — cet objet n'est jamais appelé depuis un contexte composable.
 */
object GameRequestMail {
    const val RECIPIENT = "contact@erwan-decoster.com"

    /** Retourne `false` si aucune application ne peut traiter l'intention (aucun client mail
     * installé) — laisse l'appelant proposer un repli plutôt qu'un tap sans effet visible. */
    fun open(
        context: Context,
        searchTerm: String? = null,
    ): Boolean =
        try {
            context.startActivity(intent(context, searchTerm))
            true
        } catch (error: ActivityNotFoundException) {
            false
        }

    private fun intent(
        context: Context,
        searchTerm: String?,
    ): Intent =
        Intent(Intent.ACTION_SENDTO).apply {
            data = Uri.parse("mailto:")
            putExtra(Intent.EXTRA_EMAIL, arrayOf(RECIPIENT))
            putExtra(Intent.EXTRA_SUBJECT, context.getString(R.string.demande_d_ajout_d_un_jeu_ca_compte))
            putExtra(Intent.EXTRA_TEXT, body(context, searchTerm))
        }

    private fun body(
        context: Context,
        searchTerm: String?,
    ): String {
        val intro =
            if (!searchTerm.isNullOrEmpty()) {
                context.getString(R.string.le_jeu_que_je_cherche_value1, searchTerm)
            } else {
                context.getString(R.string.le_jeu_que_je_souhaite_voir_ajoute)
            }
        val rulesPrompt = context.getString(R.string.regles_ou_lien_vers_les_regles_facultatif)
        return "$intro\n\n\n$rulesPrompt\n\n"
    }
}

File: apple/App/GameRequestMail.swift
import Foundation
import SwiftUI

#if canImport(UIKit)
  import UIKit
#endif

/// Doc utilisateur — demande d'ajout d'un jeu : l'app reste volontairement simple (pas de
/// formulaire ni de backend dédié), un `mailto:` pré-rempli suffit. Point d'entrée unique
/// partagé par la recherche sans résultat (`GamesTabView`) et les réglages (`SettingsView`), pour
/// que les deux chemins produisent exactement le même sujet/corps.
enum GameRequestMail {
  static let recipient = "contact@erwan-decoster.com"

  /// Doc utilisateur — tente d'ouvrir un client mail préempli ; retourne `false` si aucun n'est
  /// disponible (app Mail supprimée, aucun compte configuré…) — `canOpenURL` renvoie alors
  /// `false` plutôt que d'échouer silencieusement au moment d'`open`, ce qui laisse l'appelant
  /// proposer un repli (voir `gameRequestMailFallback`) au lieu d'un tap sans effet visible.
  @MainActor
  @discardableResult
  static func open(searchTerm: String? = nil) -> Bool {
    #if canImport(UIKit)
      guard let url = url(searchTerm: searchTerm), UIApplication.shared.canOpenURL(url) else {
        return false
      }
      UIApplication.shared.open(url)
      return true
    #else
      return false
    #endif
  }

  private static func url(searchTerm: String?) -> URL? {
    let subject: LocalizedStringResource = "Demande d'ajout d'un jeu — Qui Mène ?"
    var components = URLComponents()
    components.scheme = "mailto"
    components.path = recipient
    components.queryItems = [
      URLQueryItem(name: "subject", value: String(localized: subject)),
      URLQueryItem(name: "body", value: body(searchTerm: searchTerm)),
    ]
    return components.url
  }

  private static func body(searchTerm: String?) -> String {
    let intro: LocalizedStringResource
    if let searchTerm, !searchTerm.isEmpty {
      intro = "Le jeu que je cherche : \(searchTerm)"
    } else {
      intro = "Le jeu que je souhaite voir ajouté :"
    }
    let rulesPrompt: LocalizedStringResource = "Règles ou lien vers les règles (facultatif) :"
    return "\(String(localized: intro))\n\n\n\(String(localized: rulesPrompt))\n\n"
  }
}

extension View {
  /// Doc utilisateur — repli affiché quand `GameRequestMail.open()` échoue : montre l'adresse et
  /// permet de la copier, plutôt qu'un tap sur « Demander ce jeu » qui ne produirait aucun effet
  /// visible sur un appareil sans client mail configuré.
  func gameRequestMailFallback(isPresented: Binding<Bool>) -> some View {
    alert("Aucune messagerie configurée", isPresented: isPresented) {
      Button("Copier l'adresse") {
        #if canImport(UIKit)
          UIPasteboard.general.string = GameRequestMail.recipient
        #endif
      }
      Button("OK", role: .cancel) {}
    } message: {
      Text("Envoie ta demande à \(GameRequestMail.recipient) depuis l'application de ton choix.")
    }
  }
}
